File: bestia-zoneserver/src/test/java/net/bestia/zoneserver/environment/date/BestiaDateTest.java
package net.bestia.zoneserver.environment.date;

import static org.junit.Assert.assertEquals;

import java.time.LocalDateTime;
import java.util.Calendar;
import java.util.Date;

import org.junit.Assert;
import org.junit.Test;

import net.bestia.zoneserver.environment.date.BestiaDate;

public class BestiaDateTest {
	
	@Test
	public void getHours_correctHours() {
		
		BestiaDate d = BestiaDate.Companion.fromDate(createDate());
		int bhour = d.getHours(createNowDate());
		
		assertEquals(1, bhour);
	}
	
	@Test(expected=NullPointerException.class)
	public void fromDate_null_throws() {
		BestiaDate.Companion.fromDate(null);
	}
	
	@Test
	public void getMinutes_correctMinutes() {
		BestiaDate d = BestiaDate.Companion.fromDate(createDate());
		int bmin = d.getMinutes(createNowDate());
		
		assertEquals(9, bmin);
	}
	
	@Test
	public void getSeason_currentSeason() {
		BestiaDate d = BestiaDate.Companion.fromDate(createDate());
		Season s = d.getSeason();
		Assert.assertNotNull(s);
	}

	@Test
	public void fromDate_instance() {
		BestiaDate bd = BestiaDate.Companion.fromDate(createDate());
		Assert.assertNotNull(bd);
	}
	
	private LocalDateTime createNowDate() {
		return LocalDateTime.of(2016, 4, 14, 14, 10);
	}
	
	private Date createDate() {
		Calendar cal = Calendar.getInstance();
		cal.set(2016, 3, 12, 13, 00);
		return cal.getTime();
	}
}

File: bestia-zoneserver/src/test/kotlin/net/bestia/zoneserver/environment/date/BestiaDateTest.kt
package net.bestia.zoneserver.environment.date

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import java.time.Clock
import java.time.Instant
import java.time.ZoneId

internal class BestiaDateTest {

  @Test
  fun `monthProgress at startOfMonth is 0`() {
    val clock = Clock.fixed(Instant.parse("2019-01-01T00:00:00.000Z"), ZoneId.of("UTC"))
    val sut = BestiaDate(Instant.parse("2019-01-01T00:00:00.000Z"), clock)
    assertEquals(0f, sut.monthProgress)
  }

  @Test
  fun `monthProgress at endOfMonth is 1`() {
    val clock = Clock.fixed(Instant.parse("2019-01-31T23:59:59.000Z"), ZoneId.of("UTC"))
    val sut = BestiaDate(Instant.parse("2019-01-01T00:00:00.000Z"), clock)

    assertEquals(1f, sut.monthProgress)
  }

  @Test
  fun `monthProgress at start of next month is 0`() {
    val clock = Clock.fixed(Instant.parse("2019-02-01T00:00:01.000Z"), ZoneId.of("UTC"))
    val sut = BestiaDate(Instant.parse("2019-01-01T00:00:00.000Z"), clock)

    assertEquals(0f, sut.monthProgress)
  }

  @Test
  fun `year is 1 after 3 months`() {
    val clock = Clock.fixed(Instant.parse("2019-03-31T23:59:59.000Z"), ZoneId.of("UTC"))
    val sut = BestiaDate(Instant.parse("2019-01-01T00:00:00.000Z"), clock)

    assertEquals(1, sut.year)
  }

  @Test
  fun `year is 2 after 9 months`() {
    val clock = Clock.fixed(Instant.parse("2019-10-10T23:59:59.000Z"), ZoneId.of("UTC"))
    val sut = BestiaDate(Instant.parse("2019-01-01T00:00:00.000Z"), clock)

    assertEquals(2, sut.year)
  }

  @Test
  fun `season is fall after 3 months`() {
    val clock = Clock.fixed(Instant.parse("2019-04-01T23:59:59.000Z"), ZoneId.of("UTC"))
    val sut = BestiaDate(Instant.parse("2019-01-01T00:00:00.000Z"), clock)

    assertEquals(Season.AUTUMN, sut.season)
  }
}